File: pkg/sock/unix.go
package sock

import (
	"net"

	"github.com/complyue/hbi/pkg/errors"
	"github.com/complyue/hbi/pkg/proto"
	"github.com/golang/glog"
)

// ServeUnix listens on the specified file path (domain socket), serves each incoming connection with the
// hosting environment created from the `heFactory` function.
//
// `cb` will be called with the created `*net.UnixListener`.
//
// This func won't return until the listener is closed.
func ServeUnix(addr string, heFactory func() *proto.HostingEnv, cb func(*net.UnixListener)) (err error) {
	var raddr *net.UnixAddr
	raddr, err = net.ResolveUnixAddr("unix", addr)
	if nil != err {
		glog.Errorf("addr error: %+v", errors.RichError(err))
		return
	}
	var listener *net.UnixListener
	listener, err = net.ListenUnix("unix", raddr)
	if err != nil {
		glog.Errorf("listen error: %+v", errors.RichError(err))
		return
	}
	if cb != nil {
		cb(listener)
	}

	for {
		var conn *net.UnixConn
		conn, err = listener.AcceptUnix()
		if nil != err {
			glog.Errorf("accept error: %+v", errors.RichError(err))
			return
		}

		// todo DoS react
		he := heFactory()

		wire := NewSocketWire(conn)
		netIdent := wire.NetIdent()
		glog.V(1).Infof("New HBI connection accepted: %s", netIdent)

		proto.NewConnection(wire, he)
	}
}

// DialUnix connects to specified file path (domain socket), react with specified hosting environment.
//
// The returned posting endpoint is used to create posting conversations to send code & data to remote
// site for active communication.
//
// The returned hosting endpoint is used to obtain the current hosting conversation triggered by a
// posting conversation from remote site for passive communication.
func DialUnix(addr string, he *proto.HostingEnv) (po *proto.PostingEnd, ho *proto.HostingEnd, err error) {
	raddr, err := net.ResolveUnixAddr("Unix", addr)
	if nil != err {
		glog.Errorf("addr error: %+v", errors.RichError(err))
		return
	}
	conn, err := net.DialUnix("Unix", nil, raddr)
	if nil != err {
		glog.Errorf("conn error: %+v", errors.RichError(err))
		return
	}

	wire := NewSocketWire(conn)
	netIdent := wire.NetIdent()
	glog.V(1).Infof("New HBI connection established: %s", netIdent)

	po, ho, err = proto.NewConnection(wire, he)

	return
}

File: pkg/sock/unix.py
import asyncio
import inspect
import socket
import traceback
from typing import *

from .._details import *
from ..log import *
from ..proto import *
from .sock import SocketWire

__all__ = ["serve_unix", "dial_unix"]

logger = get_logger(__name__)


async def serve_unix(
    addr: str,
    he_factory: Callable[[], HostingEnv],
    *,
    wire_buf_high=20 * 1024 * 1024,
    wire_buf_low=6 * 1024 * 1024,
    net_opts: Optional[dict] = None,
):
    """
    serve_unix takes a factory function (which will be asked to produce one hosting
    env for each incoming HBI consumer connection), and returns `asyncio.Server` object
    representing the listening socket.

    """

    if net_opts is None:
        net_opts = {}

    loop = asyncio.get_running_loop()

    server = await loop.create_unix_server(
        lambda: SocketWire(HBIC(he_factory()), wire_buf_high, wire_buf_low),
        path=addr,
        **net_opts,
    )

    return server


async def dial_unix(
    addr: str,
    he: HostingEnv,
    *,
    wire_buf_high=20 * 1024 * 1024,
    wire_buf_low=6 * 1024 * 1024,
    net_opts: Optional[dict] = None,
) -> Tuple[PostingEnd, HostingEnd]:
    """
    dial_unix takes a hosting env, establishes HBI connection to a service at `addr`,
    returns the posting and hosting endpoints pair.

    """

    if net_opts is None:
        net_opts = {}

    loop = asyncio.get_running_loop()

    transport, wire = await loop.create_unix_connection(
        lambda: SocketWire(HBIC(he), wire_buf_high, wire_buf_low), path=addr, **net_opts
    )

    hbic = wire.hbic
    await hbic.wait_connected()

    return hbic.po, hbic.ho
